cmake/sub-build/sub-build.cmake

```diff
@@ -0,0 +1,115 @@
+# maintaining the efficiency of properly expressed builds.
+#
+####
+include_guard()
+include(utilities)
+include(sub-build/sub-build-config)
+
+####
+# Function `run_sub_build`:
+#
+# Runs a sub-build with the supplied SUB_BUILD_NAME. Output is set in PARENT_SCOPE to the name variable named by OUTPUT.
+# The targets registered and run as part of the sub-build are supplied in ARGN.
+#
+# Targets specified in ARGN **must** define top-level global targets in-order for the sub-build target to build
+# correctly. Defining a no-op global target will result in only the generate step performing any useful actions, which
+# may, in some cases, be desired.
+#
+# `SUB_BUILD_NAME`: name of the sub-build and directory to contain its build cache
+# `JOBS`: number of jobs to use in sub-build
+# `ARGN`: list of targets to run as part of this sub-build. These will be registered, and run in order.
+#####
+function(run_sub_build SUB_BUILD_NAME)
+    if (NOT DEFINED FPRIME_SUB_BUILD_TARGETS)
+        message(STATUS "[sub-build] Performing sub-build: ${SUB_BUILD_NAME}")
+        _get_call_properties()
+
+        # Run CMake as efficiently as possible
+        file(MAKE_DIRECTORY "${CMAKE_BINARY_DIR}/sub-build-${SUB_BUILD_NAME}")
+        if (CMAKE_DEBUG_OUTPUT)
+            message(STATUS "[sub-build] Generating: ${SUB_BUILD_NAME} with ${ARGN}")
+        endif ()
+        string(REPLACE ";" "\\;" TARGET_LIST_AS_STRING "${ARGN}")
+        execute_process_or_fail("[sub-build] Failed to generate: ${SUB_BUILD_NAME}"
+            "${CMAKE_COMMAND}"
+            -G "${CMAKE_GENERATOR}"
+            "${CMAKE_CURRENT_SOURCE_DIR}"
+            "-DFPRIME_SUB_BUILD_TARGETS=${TARGET_LIST_AS_STRING}"
+            "-DFPRIME_SKIP_TOOLS_VERSION_CHECK=ON"
+            "-DFPRIME_BINARY_DIR=${CMAKE_BINARY_DIR}"
+            ${CALL_PROPS}
+            RESULT_VARIABLE result
+            WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/sub-build-${SUB_BUILD_NAME}"
+        )
+
+        set(BUILD_EXTRA_ARGS)
+        # When specified add specific number of jobs to the sub build
+        if (DEFINED FPRIME_SUB_BUILD_JOBS)
+            list(APPEND BUILD_EXTRA_ARGS "--" "--jobs=${FPRIME_SUB_BUILD_JOBS}")
+        # Otherwise specify the MAKEFLAGS variable to speed-up makefile driven systems
+        else()
+            cmake_host_system_information(RESULT CPU_COUNT QUERY NUMBER_OF_PHYSICAL_CORES)
+            if(CPU_COUNT GREATER 0)
+                set(ENV{MAKEFLAGS} "--jobs=${CPU_COUNT}")
+            endif()
+        endif()
+        foreach (TARGET IN LISTS ARGN)
+            get_filename_component(TARGET_NAME "${TARGET}" NAME_WE)
+            if (CMAKE_DEBUG_OUTPUT)
+                message(STATUS "[sub-build] Executing: ${SUB_BUILD_NAME} with ${TARGET_NAME}")
+            endif ()
+            execute_process_or_fail("[sub-build] Failed to execute: ${SUB_BUILD_NAME}/${TARGET_NAME}"
```

Comment:
Would it be worth logging this regardless of DEBUD_OUTPUT? It's not a lot of output but does give an indication on progress - my laptop seems to be a little slow on executing the subbuilds

Reply:
Sure!